### PR TITLE
mds: unify daemon and tell commands

### PR DESCRIPTION
**Pull request by liewegas**
This converts the MDS to use the new unified asok/tell
infrastructure.

### CODE REVIEW

**Comment by batrick** [comment]
Changes look okay to me after a cusory look. @vshankar please have a gander as well.

**Comment by batrick** [comment]
rebase needed

**Comment by batrick** [comment]
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20191112.230706

**Comment by batrick** [comment]
````
2019-11-13T11:03:38.795 INFO:teuthology.orchestra.run.smithi133:Running:
2019-11-13T11:03:38.795 INFO:teuthology.orchestra.run.smithi133:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 900 ceph --cluster ceph --admin-daemon /var/run/ceph/ceph-mds.b.asok session evict 5924
2019-11-13T11:03:38.932 INFO:teuthology.orchestra.run.smithi133.stdout:ERROR: (22) Invalid argument
2019-11-13T11:03:38.938 INFO:tasks.cephfs.filesystem:_json_asok output: ERROR: (22) Invalid argument

2019-11-13T11:03:38.942 INFO:tasks.cephfs_test_runner:test_client_release_bug (tasks.cephfs.test_client_limits.TestClientLimits) ... ERROR
...
2019-11-13T11:04:58.694 INFO:tasks.cephfs_test_runner:======================================================================
2019-11-13T11:04:58.694 INFO:tasks.cephfs_test_runner:ERROR: test_client_release_bug (tasks.cephfs.test_client_limits.TestClientLimits)
2019-11-13T11:04:58.694 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-11-13T11:04:58.694 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2019-11-13T11:04:58.694 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20191112.230706/qa/tasks/cephfs/test_client_limits.py", line 146, in test_client_release_bug
2019-11-13T11:04:58.694 INFO:tasks.cephfs_test_runner:    self.fs.mds_asok(['session', 'evict', "%s" % mount_a_client_id])
2019-11-13T11:04:58.695 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20191112.230706/qa/tasks/cephfs/filesystem.py", line 962, in mds_asok
2019-11-13T11:04:58.695 INFO:tasks.cephfs_test_runner:    return self.json_asok(command, 'mds', mds_id, timeout=timeout)
2019-11-13T11:04:58.695 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20191112.230706/qa/tasks/cephfs/filesystem.py", line 204, in json_asok
2019-11-13T11:04:58.695 INFO:tasks.cephfs_test_runner:    return json.loads(response_data)
2019-11-13T11:04:58.695 INFO:tasks.cephfs_test_runner:  File "/usr/lib/python2.7/json/__init__.py", line 339, in loads
2019-11-13T11:04:58.695 INFO:tasks.cephfs_test_runner:    return _default_decoder.decode(s)
2019-11-13T11:04:58.695 INFO:tasks.cephfs_test_runner:  File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
2019-11-13T11:04:58.695 INFO:tasks.cephfs_test_runner:    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
2019-11-13T11:04:58.695 INFO:tasks.cephfs_test_runner:  File "/usr/lib/python2.7/json/decoder.py", line 382, in raw_decode
2019-11-13T11:04:58.696 INFO:tasks.cephfs_test_runner:    raise ValueError("No JSON object could be decoded")
2019-11-13T11:04:58.696 INFO:tasks.cephfs_test_runner:ValueError: No JSON object could be decoded
````

From: /ceph/teuthology-archive/pdonnell-2019-11-13_02:21:02-fs-wip-pdonnell-testing-20191112.230706-distro-basic-smithi/4501587/teuthology.log

Looks like a bug from this PR.

**Comment by batrick** [comment]
At least a few of the other failures should be fixed by #31620

**Comment by liewegas** [comment]
@batrick Ok teh problem with 'session evict' is the tell version takes a SessionFilter, which wants a ``=`` character and whatever else, and the admin socket version is just the client_id.  So... I think we need to decide which behavior we want.  I think we could just make a bare ID to be a legal "filter" and paper over this.

Otherwise we need to make changes to behavior and adjust the callers (pretty much all QA I'm guessing) to unify this... which might also be a good time/opportunity to get rid of all the duplicated commans (``session evict`` vs ``client evict``, ``client ls`` ``session ls``, ``session kill``, etc.)...

**Comment by batrick** [comment]
> @batrick Ok teh problem with 'session evict' is the tell version takes a SessionFilter, which wants a `=` character and whatever else, and the admin socket version is just the client_id. So... I think we need to decide which behavior we want. I think we could just make a bare ID to be a legal "filter" and paper over this.

bare ID as a legal filter seems reasonable to me.

> Otherwise we need to make changes to behavior and adjust the callers (pretty much all QA I'm guessing) to unify this... which might also be a good time/opportunity to get rid of all the duplicated commans (`session evict` vs `client evict`, `client ls` `session ls`, `session kill`, etc.)...

Making this change is difficult as it might break tests against older clusters during upgrades (although unlikely).

**Comment by liewegas** [comment]
> > @batrick Ok teh problem with 'session evict' is the tell version takes a SessionFilter, which wants a `=` character and whatever else, and the admin socket version is just the client_id. So... I think we need to decide which behavior we want. I think we could just make a bare ID to be a legal "filter" and paper over this.
> 
> bare ID as a legal filter seems reasonable to me.

Sounds good to me.

> > Otherwise we need to make changes to behavior and adjust the callers (pretty much all QA I'm guessing) to unify this... which might also be a good time/opportunity to get rid of all the duplicated commans (`session evict` vs `client evict`, `client ls` `session ls`, `session kill`, etc.)...
> 
> Making this change is difficult as it might break tests against older clusters during upgrades (although unlikely).

Perhaps we can at least switch all the tests in master to use either the ``session ...`` or ``client ...`` variants?  What's your preference?

**Comment by batrick** [comment]
> > > Otherwise we need to make changes to behavior and adjust the callers (pretty much all QA I'm guessing) to unify this... which might also be a good time/opportunity to get rid of all the duplicated commans (`session evict` vs `client evict`, `client ls` `session ls`, `session kill`, etc.)...
> > 
> > 
> > Making this change is difficult as it might break tests against older clusters during upgrades (although unlikely).
> 
> Perhaps we can at least switch all the tests in master to use either the `session ...` or `client ...` variants? What's your preference?

I mean if the QA code in master tries to evict a client on an older cluster (like Mimic/Nautilus), it won't work. I don't think we do that but it's something to think about. We can try to make the change and then see if anything breaks. You can update it to whichever form you think is more intuitive.

**Comment by liewegas** [comment]
Ok, this is ready to retest.  I'll save the command consolidation for a separate PR

**Comment by batrick** [comment]
Tests in 3h: http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20191205.235703

File: src/mds/MDSRank.cc [hunk 1]
```diff
@@ -2593,133 +2659,36 @@ int MDSRankDispatcher::handle_asok_command(std::string_view command,
...
-    } else {
-      ceph_abort();
-    }
-  }
-
-  void finish(int r) override {
-    f.open_object_section("result");
-    f.dump_int("return_code", r);
-    f.close_section();
-    C_ExecAndReply::finish(r);
```
Comment (batrick) [inline src/mds/MDSRank.cc]: `scrub abort` / `scrub pause` commands are missing this JSON output:

/ceph/teuthology-archive/pdonnell-2019-12-06_03:07:34-fs-wip-pdonnell-testing-20191205.235703-distro-basic-smithi/4573024/teuthology.log

File: src/mds/MDSRank.cc [hunk 2]
```diff
@@ -2593,133 +2659,36 @@ int MDSRankDispatcher::handle_asok_command(std::string_view command,
     command_openfiles_ls(f);
   } else if (command == "dump inode") {
     command_dump_inode(f, cmdmap, ss);
+  } else if (command == "damange ls") {
+    std::lock_guard l(mds_lock);
+    damage_table.dump(f);
+  } else if (command == "damange rm") {
```
Comment (batrick) [inline src/mds/MDSRank.cc]: typo

File: src/mds/MDSRank.cc [hunk 3]
```diff
@@ -2593,133 +2659,36 @@ int MDSRankDispatcher::handle_asok_command(std::string_view command,
     command_openfiles_ls(f);
   } else if (command == "dump inode") {
     command_dump_inode(f, cmdmap, ss);
+  } else if (command == "damange ls") {
```
Comment (batrick) [inline src/mds/MDSRank.cc]: typo

**Comment by liewegas** [comment]
updated!

**Comment by batrick** [comment]
@liewegas 

> Which indicates that connections are left open somehow. Can you tell how?

What was the issue?

**Comment by liewegas** [comment]
> @liewegas
> 
> > Which indicates that connections are left open somehow. Can you tell how?
> 
> What was the issue?

Oh nm, i didn't look at that issue yet.  Fixed the other two.

**Comment by liewegas** [comment]
@batrick ok fixed the con leak too!

**Comment by batrick** [comment]
In 3h: http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20191210.162125

File: src/mds/MDSDaemon.cc [hunk 4]
```diff
@@ -249,20 +331,53 @@ void MDSDaemon::set_up_admin_socket()
...
 				     "name=value,type=CephString,req=false ",
 				     asok_hook,
 				     "Config a CephFS client session");
   assert(r == 0);
+  r = admin_socket->register_command("damage ls",
+				     asok_hook,
+				     "List detected metadata damage");
+  assert(r == 0);
+  r = admin_socket->register_command("damage rm "
+				     "rm name=damage_id,type=CephInt",
```
Comment (batrick) [inline src/mds/MDSDaemon.cc]: extra "rm"

````
2019-12-10T23:00:50.962 INFO:teuthology.orchestra.run.smithi175.stderr:no valid command found; 1 closest matches:
2019-12-10T23:00:50.962 INFO:teuthology.orchestra.run.smithi175.stderr:damage rm rm <damage_id:int>
2019-12-10T23:00:50.962 INFO:teuthology.orchestra.run.smithi175.stderr:Error EINVAL: invalid command
````

Comment (liewegas) [reply src/mds/MDSDaemon.cc]: sigh...

**Comment by batrick** [comment]
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20191211.205606

File: src/mds/MDSDaemon.cc [hunk 5]
```diff
@@ -249,20 +331,53 @@ void MDSDaemon::set_up_admin_socket()
...
 				     "name=value,type=CephString,req=false ",
 				     asok_hook,
 				     "Config a CephFS client session");
   assert(r == 0);
+  r = admin_socket->register_command("damage ls",
+				     asok_hook,
+				     "List detected metadata damage");
+  assert(r == 0);
+  r = admin_socket->register_command("damage rm "
+				     " name=damage_id,type=CephInt",
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/mds/MDSDaemon.cc]: ha, indeed.  fixed

**Comment by liewegas** [comment]
\o/

**Pull request merged**